velox/functions/prestosql/window/NthValue.cpp

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "velox/common/base/Exceptions.h"
+#include "velox/exec/WindowFunction.h"
+#include "velox/expression/FunctionSignature.h"
+#include "velox/functions/prestosql/window/WindowFunctionNames.h"
+#include "velox/vector/FlatVector.h"
+
+namespace facebook::velox::window {
+
+namespace {
+
+template <typename T>
+class NthValueFunction : public exec::WindowFunction {
+ public:
+  explicit NthValueFunction(
+      const std::vector<exec::RowColumn>& argColumns,
+      const TypePtr& resultType)
+      : WindowFunction(resultType),
+        argColumns_(argColumns),
+        partitionRows_(nullptr) {}
+
+  void resetPartition(const std::vector<char*>& rows) {
+    partitionRows_ = reinterpret_cast<char* const*>(rows.data());
+    /*firstArgVector_->resize(rows.size());
+    exec::RowContainer::extractColumn(
+      rows.data(), rows.size(), argColumns_[0], firstArgVector_);
+
+    offsetsVector_->resize(rows.size());
+    exec::RowContainer::extractColumn(
+      rows.data(), rows.size(), argColumns_[1], offsetsVector_); */
+  }
+
+  void apply(
+      const BufferPtr& peerGroupStarts,
+      const BufferPtr& /*peerGroupEnds*/,
+      const BufferPtr& frameStarts,
+      const BufferPtr& /*frameEnds*/,
+      int32_t startPartitionRow,
+      int32_t resultIndex,
+      const VectorPtr& result) {
+    FlatVector<T>* resultVector = result->as<FlatVector<T>>();
+    auto firstArgVector =
+        BaseVector::create(resultVector->type(), 1, peerGroupStarts->pool());
+    auto offsetsVector =
+        BaseVector::create(BIGINT(), 1, peerGroupStarts->pool());
+
+    int numRows = peerGroupStarts->size();
+    for (int i = 0; i < numRows; i++) {
+      exec::RowContainer::extractColumn(
+          partitionRows_ + startPartitionRow + i,
+          1,
+          argColumns_[1],
+          offsetsVector);
+      const int64_t offset = offsetsVector->values()->as<int64_t>()[0];
+      VELOX_CHECK_GE(offset, 1);
+      const int64_t frameStart = frameStarts->as<size_t>()[i];
+      VELOX_CHECK_GE(frameStart, 0);
+      exec::RowContainer::extractColumn(
+          partitionRows_ + frameStart + offset - 1,
+          1,
+          argColumns_[0],
+          firstArgVector);
+      resultVector->mutableRawValues()[resultIndex + i] =
```

Comment:
This won't work for strings as we need to copy not only StringView, but the underlying string as well.